python/ray/serve/batching.py

```diff
@@ -137,7 +145,11 @@ async def wait_for_batch(self) -> List[Any]:
             # Add all new arrivals to the batch.
             while len(batch) < max_batch_size and not self.queue.empty():
                 batch.append(self.queue.get_nowait())
-            self.queue_put_event.clear()
+
+            # Only clear the put event if the queue is empty. If it's not empty
+            # we can start construcing a new batch immediately in the next loop.
+            if self.queue.empty():
+                self.queue_put_event.clear()
```

Comment:
@edoakes @alexeykudinkin I discovered a concurrency bug and added a fix in this PR. Theses are the relevant lines.

We used to clear the `queue_put_event` whenever a batch was constructed. This meant that even if the request queue had enough requests to create another batch immediately, the `batch_handler` would wait until another request was sent before constructing a batch. If no request is sent, the existing requests wait indefinitely.

Now, we clear the event only if the queue is empty.

Reply:
We discussed offline; there's no race condition here because the code that puts objects on the queue runs in the same event loop as this code. That means requests can't be added to queue between the loop and the empty check.

I added this context to the comment.